### PR TITLE
Update pipeline word heuristic to work with whitespace in token offsets

### LINKED ISSUE #18111: Word offsets of some fast tokenizers are not compatible with token classification pipeline label aggregation
### System Info

- `transformers` version: 4.21.0.dev0
- Platform: macOS-12.4-x86_64-i386-64bit
- Python version: 3.9.13
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.11.0 (False)
- Tensorflow version (GPU?): 2.9.1 (False)
- Flax version (CPU?/GPU?/TPU?): 0.5.2 (cpu)
- Jax version: 0.3.6
- JaxLib version: 0.3.5
- Using GPU in script?: N
- Using distributed or parallel set-up in script?: N

### Who can help?

Tagging @Narsil for pipelines and @SaulLu for tokenization. Let me know if I should tag anyone for specific models, but it's not really a model issue, except in terms of tokenization.

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I noticed this issue with a DeBERTa model, but it also affects some others. The high level issue is that some tokenizers include leading spaces in the offset indices, some exclude them, and some are configurable with `trim_offsets`. When offsets include leading spaces (equivalent to `trim_offsets==False`), the pipeline [word heuristic](https://github.com/huggingface/transformers/blob/afe5d42d8d1d80af911ed980c2936bfe887078f6/src/transformers/pipelines/token_classification.py#L294) doesn't work. The result is aggregating all tokens in the sequence to one label. Simple example:

```python
model_name = "brandon25/deberta-base-finetuned-ner"
model = AutoModelForTokenClassification.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

ner_aggregate = pipeline("ner", model=model, tokenizer=tokenizer, ignore_labels=[], aggregation_strategy="max")
ner_aggregate("We're from New York")
```
Result:
```
[{'entity_group': 'O', 'score': 0.9999778, 'word': " We're from New York", 'start': 0, 'end': 19}]
```

### Expected behavior

Expected result, something like:
```
[{'entity_group': 'O', 'score': 0.9999778, 'word': " We're from", 'start': 0, 'end': 10}, {'entity_group': 'O', 'score': 0.9xxx, 'word': "New York", 'start': 11, 'end': 19}]
```

If you'd like to see actual output, here's a [colab notebook with relevant models](https://colab.research.google.com/drive/1bcWotnqSPNIuAaRNkELKmKiLQheudHu1?usp=sharing) for comparison.

This affects at least these:
- DeBERTa V1
- DeBERTa V2/3
- GPT2 (tested because `DebertaTokenizerFast` is a subclass of `GPT2TokenizerFast`)
- Depending on config, Roberta (and any other tokenizer that honors `trim_offsets==False`)

The easiest solution would be to update the heuristic. [Here is a change](https://github.com/davidbenton/transformers/commit/5c43c63d401f80818d95e9cafb627607680f4dff) that works for preceding space in sequence (like current heuristic) _or_ leading space in token. I can turn into a PR if desired.

I know a lot of the default configuration matches reference implementations or published research, so I'm not sure where inconsistencies between tokenizers are desired behavior. I did notice, for example, that some sentencepiece tokenizers include leading spaces in offset indices (DeBERTa V2/3), and some don't (Albert, XLNet). I looked at the converter config and the rust code (which is pretty opaque to me), but it's not obvious to me why the offsets are different. Do you know, @SaulLu? Is that expected?

I am comparing different architectures to replace a production Bert model and was evaluating models fine tuned on an internal dataset when I ran into this. I have my manager's blessing to spend some time on this (and already have! 😂), so I'm happy to work on a PR or help out how I can.

### PR DESCRIPTION
**Pull request by davidbenton**
# What does this PR do?

This change checks for whitespace in the input string at either the
character preceding the token or in the first character of the token.
This works with tokenizers that return offsets excluding whitespace
between words or with offsets including whitespace.

Fixes #18111

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/pipelines/test_pipelines_token_classification.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_token_classification.py:542]: This one is great it's your original issue, no need to touch it.

File: tests/pipelines/test_pipelines_token_classification.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_token_classification.py:599]: I think if we can use `hf-internal-testing/tiny-random-deberta-v2`  it'd be a much smaller model which are more targeted fast tests.

It should use the same tokenizer and expose the same things.

File: tests/pipelines/test_pipelines_token_classification.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_token_classification.py:626]: Great small tests.

I seem removing your modification didn't lead to this test failing. I may have gotten it wrong, I will try again tomorrow.

I think we should add something like

```python
 self.assertEqual(
            [entity["word"] for entity in pre_entities],
            # play is NOT a subword of `"I" `
            ["I", " play", " the" " h", "armonica"],        )
```
This is mostfly for the test reader so he can reason about the `is_subword` values without having to dive into the code.

**Review by Narsil — APPROVED** [review]
LGTM.

Adding a few NITs. Feel free to accept them or not if you don't agree there.

File: tests/pipelines/test_pipelines_token_classification.py
Comment (davidbenton) [reply tests/pipelines/test_pipelines_token_classification.py:599]: Yes, good call, will update

File: tests/pipelines/test_pipelines_token_classification.py
Comment (davidbenton) [reply tests/pipelines/test_pipelines_token_classification.py:626]: Making that clearer is a good idea, will add that

Comment (davidbenton) [reply tests/pipelines/test_pipelines_token_classification.py:626]: And I believe both tests failed before the change, but will double check as well

Comment (Narsil) [reply tests/pipelines/test_pipelines_token_classification.py:626]: Please ignore.

File: tests/pipelines/test_pipelines_token_classification.py
Comment (Narsil) [reply tests/pipelines/test_pipelines_token_classification.py:629]: Works great  :) !

For future readers that don't know:

https://www.google.com/search?q=theremin&sxsrf=ALiCzsZ9I3XIC2CMr3R2TlC-D21j_0x6rQ:1659377140631&source=lnms&tbm=isch&sa=X&ved=2ahUKEwi9z4_Inab5AhURXvEDHV3RDf8Q_AUoAXoECAIQAw&biw=1918&bih=973&dpr=1

File: tests/pipelines/test_pipelines_token_classification.py
Comment (Narsil) [reply tests/pipelines/test_pipelines_token_classification.py]: Everything works. I probably made a mistake like forgetting to save or something earlier.

**Comment by Narsil** [comment]
The tests failures seem to have nothing to do with this PR.

@LysandreJik @ydshieh  maybe ?

**Comment by davidbenton** [comment]
@LysandreJik @ydshieh @Narsil I think the test failures are because I set up CircleCI to track my fork, just to see if the tests would pass there. I didn't expect it to show here on the upstream project PR, but I think that might be what we're seeing. I've disabled that for any future commits.

I'm guessing failures on a hosted, free CircleCI project are expected, right? Sorry for the CI spam.

**Comment by ydshieh** [comment]
> I'm guessing failures on a hosted, free CircleCI project are expected, right? Sorry for the CI spam.

Probably yes: I saw you have `Docker / [Docker Medium]` while `transformers` uses `Docker / [Docker X-Large]`.

Now we have to make the tests run under Hugging Face's CircleCI plan 😄

**Comment by davidbenton** [comment]
Ugh, so that stopped the HF circleci from firing? I can add a "maybe this time" commit, as is traditional with CI workflows...

**Comment by Narsil** [comment]
Ok all that's left before merging is a final approval from a core maintainer.

@sgugger maybe ?

**Review by sgugger — APPROVED** [review]
Thanks for your PR!

**Pull request merged**
